qcodes/utils/logger.py

```diff
@@ -0,0 +1,439 @@
+    global console_handler
+    global file_handler
+
+    # set loggers to the supplied levels
+    for name, level in qc.config.logger.logger_levels.items():
+        logging.getLogger(name).setLevel(level)
+
+    root_logger = logging.getLogger()
+    root_logger.setLevel(logging.DEBUG)
+
+    # remove previously set handlers
+    for handler in (console_handler, file_handler):
+        if handler is not None:
+            handler.close()
+            root_logger.removeHandler(handler)
+
+    # add qcodes handlers
+
+    format_string = LOGGING_SEPARATOR.join(FORMAT_STRING_ITEMS)
+    formatter = logging.Formatter(format_string)
+
+    # console
+    console_handler = logging.StreamHandler()
+    console_handler.setLevel(qc.config.logger.console_level)
+    console_handler.setFormatter(formatter)
+    root_logger.addHandler(console_handler)
+
+    # file
+    filename = get_log_file_name()
+    os.makedirs(os.path.dirname(filename), exist_ok=True)
+
+    file_handler = logging.handlers.TimedRotatingFileHandler(filename,
+                                                             when='midnight')
+    file_handler.setLevel(qc.config.logger.file_level)
+    file_handler.setFormatter(formatter)
+    root_logger.addHandler(file_handler)
+
+    # capture any warnings from the warnings module
+    logging.captureWarnings(capture=True)
+
+    log.info("QCoDes logger setup completed")
+
+
+def start_command_history_logger(log_dir: Optional[str]=None) -> None:
+    """
+    logging of the history of the interactive command shell
+    works only with ipython. Call function again to set new path to log file.
+
+    Args:
+        log_dir: directory where log shall be stored to. If left out defaults
+            to '~/.qcodes/logs/command_history.log'
+    """
+    from IPython import get_ipython
+    ipython = get_ipython()
+    if ipython is None:
+        log.warn("Command history can't be saved outside of IPython/jupyter")
+        return
+
+    log_dir = log_dir or os.path.join(_get_qcodes_user_path(), LOGGING_DIR)
+    filename = os.path.join(log_dir, HISTORY_LOG_NAME)
+    os.makedirs(os.path.dirname(filename), exist_ok=True)
+
+    ipython.magic("%logstop")
+    ipython.magic("%logstart -t -o {} {}".format(filename, "append"))
+    log.info("Started logging IPython history")
+
+
+def start_all_logging() -> None:
+    """
+    Starts python log module logging and ipython comand history logging.
+    """
+    start_logger()
+    start_command_history_logger()
+
+
+def log_to_dataframe(log: List[str]=None,
+                     columns: Optional[List[str]]=None,
+                     separator: Optional[str]=None) -> pandas.DataFrame:
+    """
+    Return the provided or default log string as a pandas DataFrame.
+
+    Unless `qcodes.utils.logger.LOGGING_SEPARATOR` or
+    `qcodes.utils.logger.FORMAT_STRING...` have been changed using the default
+    for the columns and separtor arguments is encouraged.
+
+    Args:
+        log: log content
```

Comment:
of course, but I didn't mean to mention it as a type but rather than as a "structure of" that log content. But the type annotation is actually enough here.